### ISSUE TITLE
Exclude specific calls from being intercepted/retried -- HeaderSanitizerContract proposal

### ISSUE THREAD

**Issue by fnicastri** [CodingAleCR/http_interceptor#112]
**Is your feature request related to a problem? Please describe.**
Looking for a way to exclude interceptors and/or the retry policy I ended up using some custom headers (mainly necessary for the retry policy)
Es: **DO_NOT_REPLY** or **UNAUTHENTICATED**
But using them on a Flutter Web project ended up giving some CORS errors due to these custom headers.

**Describe the solution you'd like**
Added a HeaderSanitizerContract to remove custom utility headers before making the request

**Describe alternatives you've considered**
Another approach could be to add specific arguments to every method to exclude the specific call from being retried/intercepted

@CodingAleCR are you interested in a pull request?

https://github.com/fnicastri/http_interceptor/tree/main

Thanks,
Francesco

**Comment by fnicastri** [comment]
nothing? @CodingAleCR

**Comment by CodingAleCR** [comment]
Hey, thanks for opening an issue with the feature request, sorry it has taken a while. It's an interesting idea this feature and it might be easy to implement. I'll take a closer look as soon as I can

**Comment by fnicastri** [comment]
> Hey, thanks for opening an issue with the feature request, sorry it has taken a while. It's an interesting idea this feature and it might be easy to implement. I'll take a closer look as soon as I can

I already have an implementation, you can see it in my fork at https://github.com/fnicastri/http_interceptor/tree/main

**Comment by CodingAleCR** [comment]
Oh, that's great, I'll take a look and see your implementation, hopefully it matches what I was thinking on doing and accept a PR on it.

**Comment by fnicastri** [comment]
any news?

**Comment by fnicastri** [comment]
🤷‍♂️

**Comment by CodingAleCR** [comment]
Hey, sorry for the delayed answer. I looked at your implementation and had my doubts about it, mostly because it could block all requests from being intercepted. 

Instead, I would prefer to go on with the alternative of adding two methods to `InterceptorContract` called: `Future<bool> shouldInterceptRequest()` and `Future<bool> shouldInterceptResponse()`, both methods have a default implementation in case you are using `extend` and will be required if you are using `implement` on your different interceptors. This also means that you can optionally apply one interceptor and ignore the others or vice-versa.

Now regarding the `RetryPolicy`, there's a built-in `bool shouldAttemptRetryOnException(Exception reason, BaseRequest request)` and `Future<bool> shouldAttemptRetryOnResponse(BaseResponse response)` that you can override to avoid retrying a given request. Maybe you could elaborate a bit more just so I can understand why it might not be enough for your needs at the moment?

The implementation would look something like the following:

**InterceptorContract**
```dart
abstract class InterceptorContract {
  Future<bool> shouldInterceptRequest() async => true;

  Future<BaseRequest> interceptRequest({required BaseRequest request});

  Future<bool> shouldInterceptResponse() async => true;

  Future<BaseResponse> interceptResponse({required BaseResponse response});
}
```

  **Logger Interceptor Example:**
```dart
class LoggerInterceptor extends InterceptorContract {
  @override
  Future<BaseRequest> interceptRequest({
    required BaseRequest request,
  }) async {
    log('----- Request -----');
    log(request.toString());
    log(request.headers.toString());
    return request;
  }

  @override
  Future<BaseResponse> interceptResponse({
    required BaseResponse response,
  }) async {
    log('----- Response -----');
    log('Code: ${response.statusCode}');
    log(response.toString());
    return response;
  }
}
```

  **Custom `shouldInterceptRequest` Interceptor Example:**
```dart
class CustomRequestCheckInterceptor extends InterceptorContract {

  @override
  Future<bool> shouldInterceptRequest() async {
    if (someCondition) {
      return false;
    }

    return true;
  }

  @override
  Future<BaseRequest> interceptRequest({
    required BaseRequest request,
  }) async {
    log('----- Request -----');
    log(request.toString());
    log(request.headers.toString());
    return request;
  }

  @override
  Future<BaseResponse> interceptResponse({
    required BaseResponse response,
  }) async {
    log('----- Response -----');
    log('Code: ${response.statusCode}');
    log(response.toString());
    return response;
  }
}
```

**Comment by CodingAleCR** [comment]
This should be available in version `2.0.0-beta.6` which is just released 🎉
